### DISCUSSION TITLE
v0.3.1: Foundation

### DISCUSSION THREAD

**Discussion by Gipson62** [atlas77-lang/atlas77#57] [Announcements]
# Atlas77 v0.3.1 "Foundation" Release:

### Run command:
- `atlas-test.exe run <PATH>`
### Stdio functions:
- `print(val)`, `read_i64()`, `read_f64()`, `read_str()`
### More types support:
- Support for `String` and `List[T]` (including indexing and concatenation)
### Match expression:
```hs
match n
  | case_1 => expression,
  | case_2 => another_expression,
  \ _ => default_expression
```
> Note: Use `\` for the last case, `_` for the default case (only in the last case), and the last case doesn't have to be a default case.

## Later plan:
We have decided to stop adding new features to the Tree Walker interpreter as it does not align with the ultimate goal of this project. Instead, we will focus on refining the Parser and Lexer by introducing a V2 for both, revamping the Visitor, and introducing a new HLIR (High-Level Intermediate Representation). The HLIR will perform multiple passes for checks and optimization, following a nano pass compiler design. Each pass will be small, but collectively contribute to building a powerful and efficient language.

Stay tuned for more exciting updates! 🚀

<hr /><em>This discussion was created from the release <a href='https://github.com/RedGear-Studio/atlas-core/releases/tag/v0.3.1'>v0.3.1: Foundation</a>.</em>

**Comment by Gipson62** [comment]
Yo, I might take the syntax of this language and revamp it. But instead of targeting a visitor pattern, I'll target the vm I'm currently making.

**Comment by ElhamAryanpur** [comment]
@Gipson62 can you write a Hello World example?

**Reply by Gipson62** [reply]
The old grammar worked like that (tho I might change it a bit later for ease of use):
```rs
let main: () -> i64 = print("hello world")

let fib: (n: i64) -> i64 = 
  match n 
  | 1 => 1,
  | 0 => 0,
  \ _ => fib(n - 1) + fib(n - 2)

let fib2: (n: i64) -> i64 =
  if n < 2
    n
  else
    fib2(n - 1) + fib2(n - 2)
```

**Reply by Gipson62** [reply]
> @Gipson62 can you write a Hello World example?

This would be a more accurate example of what could be. I don't actually need the language to be functional, it'd be nice, but it's not mandatory in any way.
```ts

function main() {
  let i: i64 = 0;
  while i < 10 {
    print(i); 
// the print could actually be a function doing an `extern_call` if `std::io` is imported. And so we'd remove the need for print instruction. So it could also manage more stuff than just print (formatting and shit)
    i = i + 1;
  }
  let fib_res: i64 = fib(25);
  print(fib_res);
}

function fib(n: i64) -> i64 {
  match n 
    | 0 => 0,
    | 1 => 1,
    \ _ => fib(n - 1) + fib(n - 2)  
}

function fib2(n: i64) -> i64 {
  if n < 2 then
    n
  else
    fib2(n - 1) + fib(n - 2)
}
```
